<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Beginner Microbiome Analysis 2025</title>
  <meta name="description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Beginner Microbiome Analysis 2025" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://cbw-dev.github.io/bookdown-template/img/bioinformatics_logo.png" />
  <meta property="og:description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="github-repo" content="cbw-dev/BMB_2025" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Beginner Microbiome Analysis 2025" />
  
  <meta name="twitter:description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="twitter:image" content="https://cbw-dev.github.io/bookdown-template/img/bioinformatics_logo.png" />




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />

<link rel="next" href="module-1.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>

 <li><a href="./"><strong><center>Beginner Microbiome Analysis</center></strong><br>
   <img src="./img/CBW_BMB_Icon.png" width="70%" alt="WORKSHOP NAME logo" style="display:block; margin-left:auto; margin-right:auto; ">
 </a></li>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#awsunix-primer"><i class="fa fa-check"></i>AWS/UNIX primer</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#class-photo"><i class="fa fa-check"></i>Class Photo</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#aws-setup-instructions"><i class="fa fa-check"></i>AWS Setup Instructions</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#data-downloads"><i class="fa fa-check"></i>Data Downloads</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#amazon-machine-image"><i class="fa fa-check"></i>Amazon Machine Image</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#meet-your-faculty"><i class="fa fa-check"></i>Meet Your Faculty</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#morgan-langille"><i class="fa fa-check"></i>Morgan Langille</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#robyn-wright"><i class="fa fa-check"></i>Robyn Wright</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#ryan-j.-mclaughlin"><i class="fa fa-check"></i>Ryan J. McLaughlin</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#juan-santana"><i class="fa fa-check"></i>Juan Santana</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#tony-liu"><i class="fa fa-check"></i>Tony Liu</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#zhibin-lu"><i class="fa fa-check"></i>Zhibin Lu</a></li>
</ul></li>
</ul></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html"><i class="fa fa-check"></i>Module 1</a>
<ul>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#lecture"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#lab-introduction-to-sequencing-data-analysis"><i class="fa fa-check"></i>Lab: Introduction to sequencing data analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#introduction"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#log-into-your-aws-instance"><i class="fa fa-check"></i>1.1 Log into your AWS instance</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#creating-directories-and-moving-around-on-the-command-line"><i class="fa fa-check"></i>1.2 Creating directories and moving around on the command line</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#use-wget-to-download-files"><i class="fa fa-check"></i>1.3 Use wget to download files</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#move-these-files-to-the-directories"><i class="fa fa-check"></i>1.4 Move these files to the directories</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#zip-and-unzip-these-files"><i class="fa fa-check"></i>1.5 Zip and unzip these files</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#create-new-tar-archive-of-files"><i class="fa fa-check"></i>1.6 Create new tar archive of files</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#unzip-tar-archive"><i class="fa fa-check"></i>1.7 Unzip tar archive</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#look-at-fasta-and-fastq-files-with-less"><i class="fa fa-check"></i>1.8 Look at fasta and fastq files with less</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#installing-programs-to-the-server"><i class="fa fa-check"></i>1.9 Installing programs to the server</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#conda-environments"><i class="fa fa-check"></i>1.10 Conda environments</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#install-fastqc-and-multiqc"><i class="fa fa-check"></i>1.11 Install fastqc and multiqc</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#perform-quality-control-on-fastq-files"><i class="fa fa-check"></i>1.12 Perform quality control on fastq files</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#htop---looking-at-the-number-of-processes-we-have-available-or-running"><i class="fa fa-check"></i>1.13 htop - looking at the number of processes we have available or running</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#back-to-the-quality-control"><i class="fa fa-check"></i>1.14 Back to the quality control</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#quality-control-on-the-data-that-well-be-using"><i class="fa fa-check"></i>1.15 Quality control on the data that we’ll be using</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#install-qiime2"><i class="fa fa-check"></i>1.16 Install QIIME2</a></li>
<li class="chapter" data-level="" data-path="module-1.html"><a href="module-1.html#answers"><i class="fa fa-check"></i>Answers</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html"><i class="fa fa-check"></i>Module 2</a>
<ul>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#lecture-1"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#lab-marker-gene-profiling"><i class="fa fa-check"></i>Lab: Marker gene profiling</a>
<ul>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#introduction-1"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s"><i class="fa fa-check"></i>1. 16S</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-first-steps"><i class="fa fa-check"></i>1.1. 16S First steps</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-denoising-the-reads-into-amplicon-sequence-variants"><i class="fa fa-check"></i>1.2. 16S Denoising the reads into amplicon sequence variants</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-assign-taxonomy-to-asvs"><i class="fa fa-check"></i>1.3. 16S Assign taxonomy to ASVs</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-filtering-resultant-table"><i class="fa fa-check"></i>1.4. 16S Filtering resultant table</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-answers"><i class="fa fa-check"></i>16S Answers</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-1"><i class="fa fa-check"></i>2. 18S</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-first-steps-1"><i class="fa fa-check"></i>2.1. 18S First steps</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-denoising-the-reads-into-amplicon-sequence-variants-1"><i class="fa fa-check"></i>2.2. 18S Denoising the reads into amplicon sequence variants</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-assign-taxonomy-to-asvs-1"><i class="fa fa-check"></i>2.3. 18S Assign taxonomy to ASVs</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-filtering-resultant-table-1"><i class="fa fa-check"></i>2.4. 18S Filtering resultant table</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#s-answers-1"><i class="fa fa-check"></i>18S Answers</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#its"><i class="fa fa-check"></i>3. ITS</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#its-first-steps"><i class="fa fa-check"></i>3.1. ITS First steps</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#its-denoising-the-reads-into-amplicon-sequence-variants"><i class="fa fa-check"></i>3.2. ITS Denoising the reads into amplicon sequence variants</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#its-assign-taxonomy-to-asvs"><i class="fa fa-check"></i>3.3. ITS Assign taxonomy to ASVs</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#its-filtering-resultant-table"><i class="fa fa-check"></i>3.4. ITS Filtering resultant table</a></li>
<li class="chapter" data-level="" data-path="module-2.html"><a href="module-2.html#its-answers"><i class="fa fa-check"></i>ITS Answers</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html"><i class="fa fa-check"></i>Module 3: Microbiome statistics and visualizations</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#lecture-2"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#lab"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#table-of-contents"><i class="fa fa-check"></i>Table of Contents</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#introduction-2"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#build-tree-with-sepp-qiime-2-plugin"><i class="fa fa-check"></i>1. Build tree with SEPP QIIME 2 plugin</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#rarefaction-curves"><i class="fa fa-check"></i>2. Generate rarefaction curves</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#diversity-ordination"><i class="fa fa-check"></i>3. Calculating diversity metrics and generating ordination plots</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#stacked-bar-chart"><i class="fa fa-check"></i>4. Generate stacked bar chart of taxa relative abundances</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#ancom"><i class="fa fa-check"></i>5. Identifying differentially abundant features with ANCOM</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#core-microbiome"><i class="fa fa-check"></i>6. Identifying shared features across groups of samples</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#heatmaps"><i class="fa fa-check"></i>7. Heatmaps</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#exporting-data"><i class="fa fa-check"></i>8. Exporting data from QIIME 2 for use in other software</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#id_18s-its"><i class="fa fa-check"></i>9. Considerations for 18S and ITS data</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#answers-1"><i class="fa fa-check"></i>Answers</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html"><i class="fa fa-check"></i>Module 4 Lab: Functional prediction and additional analyses</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#lecture-3"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#lab-1"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#introduction-3"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#differential-abundance-analysis-using-maaslin3-and-aldex2"><i class="fa fa-check"></i>4.1 Differential abundance analysis using MaAsLin3 and ALDEx2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#read-output-into-rphyloseq"><i class="fa fa-check"></i>4.1.2 Read output into R/Phyloseq</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#run-maaslin3"><i class="fa fa-check"></i>4.1.3 Run MaAsLin3</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#run-aldex2"><i class="fa fa-check"></i>4.1.4 Run ALDEx2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#combine-and-plot-differential-abundance-results-from-maaslin3-and-aldex2"><i class="fa fa-check"></i>4.1.5 Combine and plot differential abundance results from MaAsLin3 and ALDEx2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#functional-prediction-with-picrust2-1"><i class="fa fa-check"></i>4.2 Functional prediction with PICRUSt2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#constructing-and-visualising-a-co-occurrence-network"><i class="fa fa-check"></i>4.3 Constructing and visualising a co-occurrence network</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#answers-2"><i class="fa fa-check"></i>Answers</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
 <center>
 <a href="https://www.cihr-irsc.gc.ca"><img src="img/sponsors/cihr.png" width=90% style="padding-bottom: 20px" alt="Canadian Institutes of Health Research logo."></a>

 <a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.webp" width=40% style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
 <a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" width=40% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Genomics logo.">
 <br><br>
 <a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
 <a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">
 <br><br>

 <li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Beginner Microbiome Analysis 2025</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Beginner Microbiome Analysis 2025</h1>
</div>
<div id="welcome" class="section level1 hasAnchor">
<h1>Welcome<a href="index.html#welcome" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Welcome to CBW’s Beginner Microbiome Analysis Workshop!</p>
<div id="awsunix-primer" class="section level2 hasAnchor">
<h2>AWS/UNIX primer<a href="index.html#awsunix-primer" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><a href="https://bioinformaticsdotca.github.io/AWS_2025">You can find the schedule and materials for the AWS/UNIX primer here.</a></p>
</div>
<div id="class-photo" class="section level2 hasAnchor">
<h2>Class Photo<a href="index.html#class-photo" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<iframe src="https://drive.google.com/file/d/1Bkmsukt06UJmlgqHGxzKcc-ScILu-7Zw/preview" width="100%" height="600" allow="autoplay">
</iframe>

</div>
<div id="course-schedule" class="section level2 hasAnchor">
<h2>Course Schedule<a href="index.html#course-schedule" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<table>
<colgroup>
<col width="16%" />
<col width="30%" />
<col width="16%" />
<col width="37%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Time (PST)</th>
<th align="center">Day 1</th>
<th align="center">Time (PST)</th>
<th align="center">Day 2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">8:30</td>
<td align="center">Arrivals &amp; Check-in</td>
<td align="center">8:30</td>
<td align="center">Arrivals</td>
</tr>
<tr class="even">
<td align="center">9:00</td>
<td align="center">Welcome (Coordinator)</td>
<td align="center">9:00</td>
<td align="center">Module 3 Lecture</td>
</tr>
<tr class="odd">
<td align="center">9:30</td>
<td align="center">Module 1 Lecture</td>
<td align="center"></td>
<td align="center"></td>
</tr>
<tr class="even">
<td align="center">10:30</td>
<td align="center">Break (30min)</td>
<td align="center">10:00</td>
<td align="center">Break (30min)</td>
</tr>
<tr class="odd">
<td align="center">11:00</td>
<td align="center">Module 1 Lab</td>
<td align="center">10:30</td>
<td align="center">Module 3 Lab</td>
</tr>
<tr class="even">
<td align="center">13:00</td>
<td align="center">Lunch (1h)</td>
<td align="center">12:30</td>
<td align="center">Class photo + Lunch (1h)</td>
</tr>
<tr class="odd">
<td align="center">14:00</td>
<td align="center">Module 2 Lecture</td>
<td align="center">13:30</td>
<td align="center">Module 4 Lecture</td>
</tr>
<tr class="even">
<td align="center">15:00</td>
<td align="center">Break (30min)</td>
<td align="center">14:30</td>
<td align="center">Break (30min)</td>
</tr>
<tr class="odd">
<td align="center">15:30</td>
<td align="center">Module 2 Lab</td>
<td align="center">15:00</td>
<td align="center">Module 4 Lab</td>
</tr>
<tr class="even">
<td align="center"></td>
<td align="center"></td>
<td align="center">17:00</td>
<td align="center">Survey and Closing Remarks</td>
</tr>
<tr class="odd">
<td align="center">17:30</td>
<td align="center">Finished</td>
<td align="center">17:30</td>
<td align="center">Finished</td>
</tr>
</tbody>
</table>

</div>
<div id="aws-setup-instructions" class="section level2 hasAnchor">
<h2>AWS Setup Instructions<a href="index.html#aws-setup-instructions" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="data-downloads" class="section level3 hasAnchor">
<h3>Data Downloads<a href="index.html#data-downloads" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Note that some of these files are quite large. If you have issues downloading them, try copying the link address, opening a new blank tab, and entering the URL into the new tab’s address bar. If that fails, contact us at <a href="mailto:info@bioinformatics.ca" class="email">info@bioinformatics.ca</a>.</p>
<p>AMB_data.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/AMB_data.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/AMB_data.tar</a><br />
amb_module1.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/amb_module1.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/amb_module1.tar</a><br />
amb_module2.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/amb_module2.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/amb_module2.tar</a><br />
amb_module3.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/amb_module3.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/amb_module3.tar</a><br />
amb_module4.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/amb_module4.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/amb_module4.tar</a><br />
AMB_notebooks.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/AMB_notebooks.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/AMB_notebooks.tar</a><br />
bmb_module1.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/bmb_module1.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/bmb_module1.tar</a><br />
bmb_module2.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/bmb_module2.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/bmb_module2.tar</a><br />
bmb_module3.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/bmb_module3.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/bmb_module3.tar</a><br />
bmb_module4.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/bmb_module4.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/bmb_module4.tar</a><br />
bmb_raw_data.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/bmb_raw_data.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/bmb_raw_data.tar</a><br />
conda.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/conda.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/conda.tar</a><br />
ref_dbs.tar: <a href="https://hpc4health.ca/cbw/2025/MIC/ref_dbs.tar" class="uri">https://hpc4health.ca/cbw/2025/MIC/ref_dbs.tar</a></p>
</div>
<div id="amazon-machine-image" class="section level3 hasAnchor">
<h3>Amazon Machine Image<a href="index.html#amazon-machine-image" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>We have made our AWS AMI (Amazon Machine Image) publicly available. To launch your own instance, follow the <a href="https://repost.aws/knowledge-center/launch-instance-custom-ami">instructions provided by Amazon on how to launch an EC2 instance from a custom Amazon Machine Image</a>. Please note that you will need an AWS account to proceed, and that you will need to upload the CourseData files yourself.</p>
<p>Here are the details of the AMI:</p>
<p>AWS Region: us-east-1 (N. Virgina)
AMI name: CBW_MIC_250522
AMI ID: ami-01d5bd950161d8bb2</p>
<p>If you want to create and activate a new AWS account, please follow <a href="https://aws.amazon.com/premiumsupport/knowledge-center/create-and-activate-aws-account/">these instructions provided by Amazon</a>.</p>

</div>
</div>
<div id="meet-your-faculty" class="section level2 hasAnchor">
<h2>Meet Your Faculty<a href="index.html#meet-your-faculty" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="morgan-langille" class="section level3 hasAnchor">
<h3>Morgan Langille<a href="index.html#morgan-langille" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="img/faculty/morgan-langille.jpeg" width="200"></p>
<blockquote>
<p>Professor<br>
Dalhousie University</p>
</blockquote>
<p>The focus of Dr. Langille’s research is to better understand human-microbial interactions and how that can be used to improve human health. This includes leveraging novel genomic technologies and developing improved bioinformatic methods to process and integrate multi-omic data to aid in biological interpretation. These discoveries will hopefully lead to novel applications for diagnosis and therapeutics.</p>
</div>
<div id="robyn-wright" class="section level3 hasAnchor">
<h3>Robyn Wright<a href="index.html#robyn-wright" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="img/faculty/robyn-wright.jpg" width="200"></p>
<blockquote>
<p>Postdoctoral Fellow<br>
Dalhousie University</p>
</blockquote>
<p>Robyn Wright is a microbial ecologist that completed her BSc in Marine Biology at the University of Plymouth (UK) before earning her PhD in Biological Sciences at the University of Warwick (UK) at the end of 2019. Her background is in marine microbial ecology and she studied the colonization and biodegradation of plastics during her PhD as well as in her first postdoc, where she moved to the School for Resources and Environmental Studies at Dalhousie University in Nova Scotia, Canada. She currently works within the lab of Dr. Morgan Langille, where her research aims to understand the roles that microbial communities play in human and ecosystem health and disease, and to assess and benchmark the methods that we use to study microbial systems. She enjoys working on diverse microbial ecology topics and several of her current research projects study the associations between the oral microbiome and mental and physical health conditions. She is also involved with maintaining and updating computational tools, including PICRUSt2 – a popular tool used for functional prediction within microbial communities – and with developing new tools, for example, for use in verifying the results that we obtain from metagenomic taxonomic classification tools.</p>
</div>
<div id="ryan-j.-mclaughlin" class="section level3 hasAnchor">
<h3>Ryan J. McLaughlin<a href="index.html#ryan-j.-mclaughlin" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="img/faculty/ryan-mclaughlin.png" width="200"></p>
<blockquote>
<p>Ph.D. Candidate<br>
University of British Columbia</p>
</blockquote>
<p>Ryan is a professional bioinformatician with over 15 years of industry experience. His overarching interest is finding ways to apply cutting-edge genomics technologies in the exploration and characterization of microbial communities. He has special interest in how culture-free methods can be used in concert to better resolve ecosystems at both the population- and community-level. His goal is to leverage data-driven solutions, ranging from proven machine-learning techniques to emerging artificial intelligence models, to unravel the mysteries of biological life on earth and beyond. Ryan specializes in algorithm development, analysis pipeline architecture and cloud resource engineering - all culminating in the full-stack deployment of robust bioinformatics solutions. Ryan has a BS in Biophysics from The University of Scranton, an MSc in Biology (focus in bioinformatics) from Western Washington University, and is currently finishing a PhD in Bioinformatics at The University of British Columbia.</p>
</div>
<div id="juan-santana" class="section level3 hasAnchor">
<h3>Juan Santana<a href="index.html#juan-santana" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="img/faculty/juan-santana.png" width="200"></p>
<blockquote>
<p>Ph.D. Candidate<br>
University of British Columbia</p>
</blockquote>
<p>Juan Santana holds BSc degrees in Microbiology and Biology from the University of Los Andes (Colombia), where he studied Lysinibacillus sphaericus for mosquito control as both an undergraduate and research microbiologist. He later earned an MSc in Microbiology and Biotechnology from the University of Alberta, investigating microbial communities in reclaimed soils from the Athabasca Oil Sands Region and working on biochar-based land reclamation strategies. Now a PhD candidate in the Hallam Lab at the University of British Columbia, Juan applies integrated multi-omics and statistical approaches to optimize anaerobic digestion systems for renewable natural gas production and waste-to-energy conversion.</p>
</div>
<div id="tony-liu" class="section level3 hasAnchor">
<h3>Tony Liu<a href="index.html#tony-liu" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="img/faculty/tony-liu.jpg" width="200"></p>
<blockquote>
<p>Ph.D. Candidate<br>
University of British Columbia</p>
</blockquote>
<p>Tony completed his BSc at the University of British columbia studying Microbiology and Computer Science. Continuing on at the Hallam Lab, he is currently modelling microbial metabolism at different scales of biological organization in the context of mobile genetic elements and cyanobacterial communities. Recognizing the reproducibility crisis in bioinformatics, Tony is developing a workflow automation framework to integrate existing standards for software portability and interoperability. He ultimately aims to create archive-stable computational pipelines for modelling the installation of functional gene modules into genome operating systems.</p>
</div>
<div id="zhibin-lu" class="section level3 hasAnchor">
<h3>Zhibin Lu<a href="index.html#zhibin-lu" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><img src="img/faculty/zhibin-lu.jpeg" width="200"></p>
<blockquote>
<p>HPC and Bioinformatics Services Manager<br>
Princess Margaret Cancer Centre, University Health Network</p>
</blockquote>
<p>Zhibin Lu is a senior manager at University Health Network Digital. He is responsible for UHN
HPC operations and scientific software. He manages two HPC clusters at UHN, including
system administration, user management, and maintenance of bioinformatics tools for
HPC4health. He is also skilled in Next-Gen sequence data analysis and has developed and
maintained bioinformatics pipelines at the Bioinformatics and HPC Core. He is a member of the
Digital Research Alliance of Canada Bioinformatics National Team and Scheduling National
Team.</p>

</div>
</div>
</div>



            </section>

          </div>
        </div>
      </div>

<a href="module-1.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/cbw-dev/bookdown-template/blob/main/index.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

</body>

</html>
