<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Beginner Microbiome Analysis 2025</title>
  <meta name="description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Beginner Microbiome Analysis 2025" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://cbw-dev.github.io/bookdown-template/img/bioinformatics_logo.png" />
  <meta property="og:description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="github-repo" content="cbw-dev/BMB_2025" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Beginner Microbiome Analysis 2025" />
  
  <meta name="twitter:description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="twitter:image" content="https://cbw-dev.github.io/bookdown-template/img/bioinformatics_logo.png" />




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />

<link rel="next" href="module-1-lab-introduction-to-sequencing-data-analysis.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>

 <li><a href="./">Beginner Microbiome Analysis</a></li>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#awsunix-primer"><i class="fa fa-check"></i>AWS/UNIX primer</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#meet-your-faculty"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
</ul></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html"><i class="fa fa-check"></i>Module 1 Lab: Introduction to sequencing data analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#introduction"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#log-into-your-aws-instance"><i class="fa fa-check"></i>1.1 Log into your AWS instance</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#maclinux"><i class="fa fa-check"></i>Mac/Linux</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#windows"><i class="fa fa-check"></i>Windows</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#creating-directories-and-moving-around-on-the-command-line"><i class="fa fa-check"></i>1.2 Creating directories and moving around on the command line</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#use-wget-to-download-files"><i class="fa fa-check"></i>1.3 Use wget to download files</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#move-these-files-to-the-directories"><i class="fa fa-check"></i>1.4 Move these files to the directories</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#zip-and-unzip-these-files"><i class="fa fa-check"></i>1.5 Zip and unzip these files</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#create-new-tar-archive-of-files"><i class="fa fa-check"></i>1.6 Create new tar archive of files</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#unzip-tar-archive"><i class="fa fa-check"></i>1.7 Unzip tar archive</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#look-at-fasta-and-fastq-files-with-less"><i class="fa fa-check"></i>1.8 Look at fasta and fastq files with less</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#installing-programs-to-the-server"><i class="fa fa-check"></i>1.9 Installing programs to the server</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#conda-environments"><i class="fa fa-check"></i>1.10 Conda environments</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#install-fastqc-and-multiqc"><i class="fa fa-check"></i>1.11 Install fastqc and multiqc</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#perform-quality-control-on-fastq-files"><i class="fa fa-check"></i>1.12 Perform quality control on fastq files</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#htop---looking-at-the-number-of-processes-we-have-available-or-running"><i class="fa fa-check"></i>1.13 htop - looking at the number of processes we have available or running</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#back-to-the-quality-control"><i class="fa fa-check"></i>1.14 Back to the quality control</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#quality-control-on-the-data-that-well-be-using"><i class="fa fa-check"></i>1.15 Quality control on the data that we’ll be using</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#install-qiime2"><i class="fa fa-check"></i>1.16 Install QIIME2</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#answers"><i class="fa fa-check"></i>Answers</a></li>
</ul></li>
<li class="part"><span><b>III Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html"><i class="fa fa-check"></i>Module 2 Lab: Marker gene profiling</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#introduction-1"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s"><i class="fa fa-check"></i>1. 16S</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-first-steps"><i class="fa fa-check"></i>1.1. 16S First steps</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#inspect-raw-data"><i class="fa fa-check"></i>1.1.1. Inspect raw data</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#import-fastqs-as-qiime2-artifact"><i class="fa fa-check"></i>1.1.2. Import FASTQs as QIIME2 artifact</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#trim-primers-with-cutadapt"><i class="fa fa-check"></i>1.1.3. Trim primers with cutadapt</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-denoising-the-reads-into-amplicon-sequence-variants"><i class="fa fa-check"></i>1.2. 16S Denoising the reads into amplicon sequence variants</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#join-paired-end-reads"><i class="fa fa-check"></i>1.2.1. Join paired-end reads</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#filter-out-low-quality-reads"><i class="fa fa-check"></i>1.2.2. Filter out low-quality reads</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#summarize-joined-and-filtered-reads"><i class="fa fa-check"></i>1.2.3. Summarize joined and filtered reads</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#running-deblur"><i class="fa fa-check"></i>1.2.4. Running Deblur</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#summarizing-deblur-output"><i class="fa fa-check"></i>1.2.5. Summarizing Deblur output</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-assign-taxonomy-to-asvs"><i class="fa fa-check"></i>1.3. 16S Assign taxonomy to ASVs</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#run-taxonomic-classification"><i class="fa fa-check"></i>1.3.1. Run taxonomic classification</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#assess-subset-of-taxonomic-assignments-with-blast"><i class="fa fa-check"></i>1.3.2 Assess subset of taxonomic assignments with BLAST</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-filtering-resultant-table"><i class="fa fa-check"></i>1.4. 16S Filtering resultant table</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#filter-out-rare-asvs"><i class="fa fa-check"></i>1.4.1. Filter out rare ASVs</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#filter-out-contaminant-and-unclassified-asvs"><i class="fa fa-check"></i>1.4.2. Filter out contaminant and unclassified ASVs</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#optional-exclude-low-depth-samples"><i class="fa fa-check"></i><em>1.4.3. (Optional) Exclude low-depth samples</em></a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#subset-and-summarize-filtered-table"><i class="fa fa-check"></i>1.4.4. Subset and summarize filtered table</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-answers"><i class="fa fa-check"></i>16S Answers</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-1"><i class="fa fa-check"></i>2. 18S</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-first-steps-1"><i class="fa fa-check"></i>2.1. 18S First steps</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#inspect-raw-data-1"><i class="fa fa-check"></i>2.1.1. Inspect raw data</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#import-fastqs-as-qiime2-artifact-1"><i class="fa fa-check"></i>2.1.2. Import FASTQs as QIIME2 artifact</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#trim-primers-with-cutadapt-1"><i class="fa fa-check"></i>2.1.3. Trim primers with cutadapt</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-denoising-the-reads-into-amplicon-sequence-variants-1"><i class="fa fa-check"></i>2.2. 18S Denoising the reads into amplicon sequence variants</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#join-paired-end-reads-1"><i class="fa fa-check"></i>2.2.1. Join paired-end reads</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#filter-out-low-quality-reads-1"><i class="fa fa-check"></i>2.2.2. Filter out low-quality reads</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#summarize-joined-and-filtered-reads-1"><i class="fa fa-check"></i>2.2.3. Summarize joined and filtered reads</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#running-deblur-1"><i class="fa fa-check"></i>2.2.4. Running Deblur</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#summarizing-deblur-output-1"><i class="fa fa-check"></i>2.2.5. Summarizing Deblur output</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-assign-taxonomy-to-asvs-1"><i class="fa fa-check"></i>2.3. 18S Assign taxonomy to ASVs</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#run-taxonomic-classification-1"><i class="fa fa-check"></i>2.3.1. Run taxonomic classification</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#assess-subset-of-taxonomic-assignments-with-blast-1"><i class="fa fa-check"></i>2.3.2 Assess subset of taxonomic assignments with BLAST</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-filtering-resultant-table-1"><i class="fa fa-check"></i>2.4. 18S Filtering resultant table</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#filter-out-rare-asvs-1"><i class="fa fa-check"></i>2.4.1. Filter out rare ASVs</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#filter-out-contaminant-and-unclassified-asvs-1"><i class="fa fa-check"></i>2.4.2. Filter out contaminant and unclassified ASVs</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#optional-exclude-low-depth-samples-1"><i class="fa fa-check"></i><em>2.4.3. (Optional) Exclude low-depth samples</em></a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#subset-and-summarize-filtered-table-1"><i class="fa fa-check"></i>2.4.4. Subset and summarize filtered table</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#s-answers-1"><i class="fa fa-check"></i>18S Answers</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#its"><i class="fa fa-check"></i>3. ITS</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#its-first-steps"><i class="fa fa-check"></i>3.1. ITS First steps</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#inspect-raw-data-2"><i class="fa fa-check"></i>3.1.1. Inspect raw data</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#import-fastqs-as-qiime2-artifact-2"><i class="fa fa-check"></i>3.1.2. Import FASTQs as QIIME2 artifact</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#trim-primers-with-cutadapt-2"><i class="fa fa-check"></i>3.1.3. Trim primers with cutadapt</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#its-denoising-the-reads-into-amplicon-sequence-variants"><i class="fa fa-check"></i>3.2. ITS Denoising the reads into amplicon sequence variants</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#join-paired-end-reads-2"><i class="fa fa-check"></i>2.1. Join paired-end reads</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#filter-out-low-quality-reads-2"><i class="fa fa-check"></i>3.2.2. Filter out low-quality reads</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#summarize-joined-and-filtered-reads-2"><i class="fa fa-check"></i>3.2.3. Summarize joined and filtered reads</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#running-deblur-2"><i class="fa fa-check"></i>3.2.4. Running Deblur</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#summarizing-deblur-output-2"><i class="fa fa-check"></i>3.2.5. Summarizing Deblur output</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#its-assign-taxonomy-to-asvs"><i class="fa fa-check"></i>3.3. ITS Assign taxonomy to ASVs</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#run-taxonomic-classification-2"><i class="fa fa-check"></i>3.3.1. Run taxonomic classification</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#assess-subset-of-taxonomic-assignments-with-blast-2"><i class="fa fa-check"></i>3.3.2. Assess subset of taxonomic assignments with BLAST</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#its-filtering-resultant-table"><i class="fa fa-check"></i>3.4. ITS Filtering resultant table</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#filter-out-rare-asvs-2"><i class="fa fa-check"></i>3.4.1. Filter out rare ASVs</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#filter-out-contaminant-and-unclassified-asvs-2"><i class="fa fa-check"></i>3.4.2. Filter out contaminant and unclassified ASVs</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#optional-exclude-low-depth-samples-2"><i class="fa fa-check"></i><em>3.4.3. (Optional) Exclude low-depth samples</em></a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#subset-and-summarize-filtered-table-2"><i class="fa fa-check"></i>3.4.4. Subset and summarize filtered table</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#its-answers"><i class="fa fa-check"></i>ITS Answers</a></li>
</ul></li>
<li class="part"><span><b>IV Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html"><i class="fa fa-check"></i>Module 3: Microbiome statistics and visualizations</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#table-of-contents"><i class="fa fa-check"></i>Table of Contents</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#introduction-2"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#build-tree-with-sepp-qiime-2-plugin"><i class="fa fa-check"></i>1. Build tree with SEPP QIIME 2 plugin</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#rarefaction-curves"><i class="fa fa-check"></i>2. Generate rarefaction curves</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#diversity-ordination"><i class="fa fa-check"></i>3. Calculating diversity metrics and generating ordination plots</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#stacked-bar-chart"><i class="fa fa-check"></i>4. Generate stacked bar chart of taxa relative abundances</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#ancom"><i class="fa fa-check"></i>5. Identifying differentially abundant features with ANCOM</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#core-microbiome"><i class="fa fa-check"></i>6. Identifying shared features across groups of samples</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#heatmaps"><i class="fa fa-check"></i>7. Heatmaps</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#exporting-data"><i class="fa fa-check"></i>8. Exporting data from QIIME 2 for use in other software</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#id_18s-its"><i class="fa fa-check"></i>9. Considerations for 18S and ITS data</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#building-a-de-novo-tree-for-18s"><i class="fa fa-check"></i>Building a de novo tree for 18S</a></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#building-the-its-tree"><i class="fa fa-check"></i>Building the ITS tree</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-microbiome-statistics-and-visualizations.html"><a href="module-3-microbiome-statistics-and-visualizations.html#answers-1"><i class="fa fa-check"></i>Answers</a></li>
</ul></li>
<li class="part"><span><b>V Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html"><i class="fa fa-check"></i>Module 4 Lab: Functional prediction and additional analyses</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#introduction-3"><i class="fa fa-check"></i>Introduction</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#differential-abundance-testing-with-maaslin3-and-aldex2"><i class="fa fa-check"></i>Differential abundance testing with MaAsLin3 and ALDEx2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#functional-prediction-with-picrust2"><i class="fa fa-check"></i>Functional prediction with PICRUSt2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#constructing-and-visualising-a-co-occurrence-network-using-spiec-easi-bonus-we-run-an-indicator-species-analysis-too"><i class="fa fa-check"></i>Constructing and visualising a co-occurrence network using <code>SPIEC-EASI</code> (Bonus: we run an indicator species analysis too!!)</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#differential-abundance-analysis-using-maaslin3-and-aldex2"><i class="fa fa-check"></i>4.1 Differential abundance analysis using MaAsLin3 and ALDEx2</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#prepare-data-from-the-end-of-module-2"><i class="fa fa-check"></i>4.1.1 Prepare data from the end of Module 2</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#read-output-into-rphyloseq"><i class="fa fa-check"></i>4.1.2 Read output into R/Phyloseq</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#importing-packages"><i class="fa fa-check"></i>Importing packages</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#run-maaslin3"><i class="fa fa-check"></i>4.1.3 Run MaAsLin3</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#run-aldex2"><i class="fa fa-check"></i>4.1.4 Run ALDEx2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#combine-and-plot-differential-abundance-results-from-maaslin3-and-aldex2"><i class="fa fa-check"></i>4.1.5 Combine and plot differential abundance results from MaAsLin3 and ALDEx2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#functional-prediction-with-picrust2-1"><i class="fa fa-check"></i>4.2 Functional prediction with PICRUSt2</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#filter-input-for-picrust2"><i class="fa fa-check"></i>4.2.1 Filter input for PICRUSt2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#start-running-picrust2"><i class="fa fa-check"></i>4.2.2 Start running PICRUSt2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#read-picrust2-output-into-rphyloseq"><i class="fa fa-check"></i>4.2.3 Read PICRUSt2 output into R/Phyloseq</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#look-at-picrust2-alpha-and-beta-diversity"><i class="fa fa-check"></i>4.2.3 Look at PICRUSt2 alpha and beta diversity</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#run-picrust2-differential-abundance"><i class="fa fa-check"></i>4.2.4 Run PICRUSt2 differential abundance</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#combine-and-plot-picrust2-differential-abundance-results-from-maaslin3-and-aldex2"><i class="fa fa-check"></i>4.2.5 Combine and plot PICRUSt2 differential abundance results from MaAsLin3 and ALDEx2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#plot-da-single-pathway"><i class="fa fa-check"></i>Plot DA single pathway</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#plot-da-all-pathways"><i class="fa fa-check"></i>Plot DA all pathways</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#constructing-and-visualising-a-co-occurrence-network"><i class="fa fa-check"></i>4.3 Constructing and visualising a co-occurrence network</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#introduction-4"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#prepare-data-from-the-end-of-module-2-1"><i class="fa fa-check"></i>4.3.1 Prepare data from the end of Module 2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#indicator-species-analysis-isa"><i class="fa fa-check"></i>4.3.2 Indicator Species Analysis (ISA)</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#abundance-and-prevalence-filtering"><i class="fa fa-check"></i>4.3.3 Abundance and Prevalence filtering</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#contructing-the-co-occurrence-network"><i class="fa fa-check"></i>4.3.4 Contructing the Co-occurrence Network</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#visualizing-networks"><i class="fa fa-check"></i>4.3.5 Visualizing Networks</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#bonus-size-by-network-statistics"><i class="fa fa-check"></i>Bonus: Size by network statistics</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#answers-2"><i class="fa fa-check"></i>Answers</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#answers-16s-differential-abundance-and-picrust2"><i class="fa fa-check"></i>Answers 16S differential abundance and PICRUSt2</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#answers-18s-differential-abundance"><i class="fa fa-check"></i>Answers 18S differential abundance</a></li>
<li class="chapter" data-level="" data-path="module-4-lab-functional-prediction-and-additional-analyses.html"><a href="module-4-lab-functional-prediction-and-additional-analyses.html#answers-its-differential-abundance"><i class="fa fa-check"></i>Answers ITS differential abundance</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
 <center>
 <a href="https://www.cihr-irsc.gc.ca"><img src="img/sponsors/cihr.png" width=90% style="padding-bottom: 20px" alt="Canadian Institutes of Health Research logo."></a>

 <a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.webp" width=40% style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
 <a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" width=40% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Genomics logo.">
 <br><br>
 <a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
 <a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">
 <br><br>

 <li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Beginner Microbiome Analysis 2025</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Beginner Microbiome Analysis 2025</h1>
</div>
<div id="welcome" class="section level1 hasAnchor">
<h1>Welcome<a href="index.html#welcome" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<!-- NOTE: Go to the README.md to help understand what to fill in for this workshop template! -->
<p>Welcome to CBW’s Beginner Microbiome Analysis Workshop!</p>
<div id="awsunix-primer" class="section level2 hasAnchor">
<h2>AWS/UNIX primer<a href="index.html#awsunix-primer" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><a href="https://bioinformaticsdotca.github.io/AWS_2025">You can find the schedule and materials for the AWS/UNIX primer here.</a></p>

</div>
<div id="course-schedule" class="section level2 hasAnchor">
<h2>Course Schedule<a href="index.html#course-schedule" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<table>
<colgroup>
<col width="16%" />
<col width="30%" />
<col width="16%" />
<col width="37%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Time (PST)</th>
<th align="center">Day 1</th>
<th align="center">Time (PST)</th>
<th align="center">Day 2</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">8:30</td>
<td align="center">Arrivals &amp; Check-in</td>
<td align="center">8:30</td>
<td align="center">Arrivals</td>
</tr>
<tr class="even">
<td align="center">9:00</td>
<td align="center">Welcome (Coordinator)</td>
<td align="center">9:00</td>
<td align="center">Module 3 Lecture</td>
</tr>
<tr class="odd">
<td align="center">9:30</td>
<td align="center">Module 1 Lecture</td>
<td align="center"></td>
<td align="center"></td>
</tr>
<tr class="even">
<td align="center">10:30</td>
<td align="center">Break (30min)</td>
<td align="center">10:00</td>
<td align="center">Break (30min)</td>
</tr>
<tr class="odd">
<td align="center">11:00</td>
<td align="center">Module 1 Lab</td>
<td align="center">10:30</td>
<td align="center">Module 3 Lab</td>
</tr>
<tr class="even">
<td align="center">13:00</td>
<td align="center">Lunch (1h)</td>
<td align="center">12:30</td>
<td align="center">Class photo + Lunch (1h)</td>
</tr>
<tr class="odd">
<td align="center">14:00</td>
<td align="center">Module 2 Lecture</td>
<td align="center">13:30</td>
<td align="center">Module 4 Lecture</td>
</tr>
<tr class="even">
<td align="center">15:00</td>
<td align="center">Break (30min)</td>
<td align="center">14:30</td>
<td align="center">Break (30min)</td>
</tr>
<tr class="odd">
<td align="center">15:30</td>
<td align="center">Module 2 Lab</td>
<td align="center">15:00</td>
<td align="center">Module 4 Lab</td>
</tr>
<tr class="even">
<td align="center"></td>
<td align="center"></td>
<td align="center">17:00</td>
<td align="center">Survey and Closing Remarks</td>
</tr>
<tr class="odd">
<td align="center">17:30</td>
<td align="center">Finished</td>
<td align="center">17:30</td>
<td align="center">Finished</td>
</tr>
</tbody>
</table>


</div>
<div id="meet-your-faculty" class="section level2 hasAnchor">
<h2>Meet Your Faculty<a href="index.html#meet-your-faculty" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Coming soon!</p>

</div>
</div>



            </section>

          </div>
        </div>
      </div>

<a href="module-1-lab-introduction-to-sequencing-data-analysis.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/cbw-dev/bookdown-template/blob/main/index.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

</body>

</html>
