<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>1. 16S | Beginner Microbiome Analysis 2025</title>
  <meta name="description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="1. 16S | Beginner Microbiome Analysis 2025" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://cbw-dev.github.io/bookdown-template/img/bioinformatics_logo.png" />
  <meta property="og:description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="github-repo" content="cbw-dev/BMB_2025" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="1. 16S | Beginner Microbiome Analysis 2025" />
  
  <meta name="twitter:description" content="CBW’s Bookdown Template for Workshops" />
  <meta name="twitter:image" content="https://cbw-dev.github.io/bookdown-template/img/bioinformatics_logo.png" />




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-2-lab-marker-gene-profiling.html"/>
<link rel="next" href="s-1.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>



<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>

 <li><a href="./">[YOUR WORKSHOP NAME]</a></li>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Welcome</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#awsunix-primer"><i class="fa fa-check"></i>AWS/UNIX primer</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#course-schedule"><i class="fa fa-check"></i>Course Schedule</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#meet-your-faculty"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
</ul></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html"><i class="fa fa-check"></i>Module 1 Lab: Introduction to sequencing data analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#table-of-contents"><i class="fa fa-check"></i>Table of Contents</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#introduction"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#log-into-your-aws-instance"><i class="fa fa-check"></i>1.1 Log into your AWS instance</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#maclinux"><i class="fa fa-check"></i>Mac/Linux</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#windows"><i class="fa fa-check"></i>Windows</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#creating-directories-and-moving-around-on-the-command-line"><i class="fa fa-check"></i>1.2 Creating directories and moving around on the command line</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#use-wget-to-download-files"><i class="fa fa-check"></i>1.3 Use wget to download files</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#move-these-files-to-the-directories"><i class="fa fa-check"></i>1.4 Move these files to the directories</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#zip-and-unzip-these-files"><i class="fa fa-check"></i>1.5 Zip and unzip these files</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#create-new-tar-archive-of-files"><i class="fa fa-check"></i>1.6 Create new tar archive of files</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#unzip-tar-archive"><i class="fa fa-check"></i>1.7 Unzip tar archive</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#look-at-fasta-and-fastq-files-with-less"><i class="fa fa-check"></i>1.8 Look at fasta and fastq files with less</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#installing-programs-to-the-server"><i class="fa fa-check"></i>1.9 Installing programs to the server</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#conda-environments"><i class="fa fa-check"></i>1.10 Conda environments</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#install-fastqc-and-multiqc"><i class="fa fa-check"></i>1.11 Install fastqc and multiqc</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#perform-quality-control-on-fastq-files"><i class="fa fa-check"></i>1.12 Perform quality control on fastq files</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#htop---looking-at-the-number-of-processes-we-have-available-or-running"><i class="fa fa-check"></i>1.13 htop - looking at the number of processes we have available or running</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#back-to-the-quality-control"><i class="fa fa-check"></i>1.14 Back to the quality control</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#quality-control-on-the-data-that-well-be-using"><i class="fa fa-check"></i>1.15 Quality control on the data that we’ll be using</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#install-qiime2"><i class="fa fa-check"></i>1.16 Install QIIME2</a></li>
<li class="chapter" data-level="" data-path="module-1-lab-introduction-to-sequencing-data-analysis.html"><a href="module-1-lab-introduction-to-sequencing-data-analysis.html#answers"><i class="fa fa-check"></i>Answers</a></li>
</ul></li>
<li class="part"><span><b>III Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html"><i class="fa fa-check"></i>Module 2 Lab: Marker gene profiling</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#introduction-1"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-2-lab-marker-gene-profiling.html"><a href="module-2-lab-marker-gene-profiling.html#table-of-contents-1"><i class="fa fa-check"></i>Table of Contents</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html"><i class="fa fa-check"></i>1. 16S</a>
<ul>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#s-first-steps"><i class="fa fa-check"></i>1.1. 16S First steps</a>
<ul>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#inspect-raw-data"><i class="fa fa-check"></i>1.1.1. Inspect raw data</a></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#import-fastqs-as-qiime2-artifact"><i class="fa fa-check"></i>1.1.2. Import FASTQs as QIIME2 artifact</a></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#trim-primers-with-cutadapt"><i class="fa fa-check"></i>1.1.3. Trim primers with cutadapt</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#s-denoising-the-reads-into-amplicon-sequence-variants"><i class="fa fa-check"></i>1.2. 16S Denoising the reads into amplicon sequence variants</a>
<ul>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#join-paired-end-reads"><i class="fa fa-check"></i>1.2.1. Join paired-end reads</a></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#filter-out-low-quality-reads"><i class="fa fa-check"></i>1.2.2. Filter out low-quality reads</a></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#summarize-joined-and-filtered-reads"><i class="fa fa-check"></i>1.2.3. Summarize joined and filtered reads</a></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#running-deblur"><i class="fa fa-check"></i>1.2.4. Running Deblur</a></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#summarizing-deblur-output"><i class="fa fa-check"></i>1.2.5. Summarizing Deblur output</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#s-assign-taxonomy-to-asvs"><i class="fa fa-check"></i>1.3. 16S Assign taxonomy to ASVs</a>
<ul>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#run-taxonomic-classification"><i class="fa fa-check"></i>1.3.1. Run taxonomic classification</a></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#assess-subset-of-taxonomic-assignments-with-blast"><i class="fa fa-check"></i>1.3.2 Assess subset of taxonomic assignments with BLAST</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#s-filtering-resultant-table"><i class="fa fa-check"></i>1.4. 16S Filtering resultant table</a>
<ul>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#filter-out-rare-asvs"><i class="fa fa-check"></i>1.4.1. Filter out rare ASVs</a></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#filter-out-contaminant-and-unclassified-asvs"><i class="fa fa-check"></i>1.4.2. Filter out contaminant and unclassified ASVs</a></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#optional-exclude-low-depth-samples"><i class="fa fa-check"></i><em>1.4.3. (Optional) Exclude low-depth samples</em></a></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#subset-and-summarize-filtered-table"><i class="fa fa-check"></i>1.4.4. Subset and summarize filtered table</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="s.html"><a href="s.html#s-answers"><i class="fa fa-check"></i>16S Answers</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html"><i class="fa fa-check"></i>2. 18S</a>
<ul>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#s-first-steps-1"><i class="fa fa-check"></i>2.1. 18S First steps</a>
<ul>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#inspect-raw-data-1"><i class="fa fa-check"></i>2.1.1. Inspect raw data</a></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#import-fastqs-as-qiime2-artifact-1"><i class="fa fa-check"></i>2.1.2. Import FASTQs as QIIME2 artifact</a></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#trim-primers-with-cutadapt-1"><i class="fa fa-check"></i>2.1.3. Trim primers with cutadapt</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#s-denoising-the-reads-into-amplicon-sequence-variants-1"><i class="fa fa-check"></i>2.2. 18S Denoising the reads into amplicon sequence variants</a>
<ul>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#join-paired-end-reads-1"><i class="fa fa-check"></i>2.2.1. Join paired-end reads</a></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#filter-out-low-quality-reads-1"><i class="fa fa-check"></i>2.2.2. Filter out low-quality reads</a></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#summarize-joined-and-filtered-reads-1"><i class="fa fa-check"></i>2.2.3. Summarize joined and filtered reads</a></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#running-deblur-1"><i class="fa fa-check"></i>2.2.4. Running Deblur</a></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#summarizing-deblur-output-1"><i class="fa fa-check"></i>2.2.5. Summarizing Deblur output</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#s-assign-taxonomy-to-asvs-1"><i class="fa fa-check"></i>2.3. 18S Assign taxonomy to ASVs</a>
<ul>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#run-taxonomic-classification-1"><i class="fa fa-check"></i>2.3.1. Run taxonomic classification</a></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#assess-subset-of-taxonomic-assignments-with-blast-1"><i class="fa fa-check"></i>2.3.2 Assess subset of taxonomic assignments with BLAST</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#s-filtering-resultant-table-1"><i class="fa fa-check"></i>2.4. 18S Filtering resultant table</a>
<ul>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#filter-out-rare-asvs-1"><i class="fa fa-check"></i>2.4.1. Filter out rare ASVs</a></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#filter-out-contaminant-and-unclassified-asvs-1"><i class="fa fa-check"></i>2.4.2. Filter out contaminant and unclassified ASVs</a></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#optional-exclude-low-depth-samples-1"><i class="fa fa-check"></i><em>2.4.3. (Optional) Exclude low-depth samples</em></a></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#subset-and-summarize-filtered-table-1"><i class="fa fa-check"></i>2.4.4. Subset and summarize filtered table</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="s-1.html"><a href="s-1.html#s-answers-1"><i class="fa fa-check"></i>18S Answers</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html"><i class="fa fa-check"></i>3. ITS</a>
<ul>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#its-first-steps"><i class="fa fa-check"></i>3.1. ITS First steps</a>
<ul>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#inspect-raw-data-2"><i class="fa fa-check"></i>3.1.1. Inspect raw data</a></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#import-fastqs-as-qiime2-artifact-2"><i class="fa fa-check"></i>3.1.2. Import FASTQs as QIIME2 artifact</a></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#trim-primers-with-cutadapt-2"><i class="fa fa-check"></i>3.1.3. Trim primers with cutadapt</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#its-denoising-the-reads-into-amplicon-sequence-variants"><i class="fa fa-check"></i>3.2. ITS Denoising the reads into amplicon sequence variants</a>
<ul>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#join-paired-end-reads-2"><i class="fa fa-check"></i>2.1. Join paired-end reads</a></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#filter-out-low-quality-reads-2"><i class="fa fa-check"></i>3.2.2. Filter out low-quality reads</a></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#summarize-joined-and-filtered-reads-2"><i class="fa fa-check"></i>3.2.3. Summarize joined and filtered reads</a></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#running-deblur-2"><i class="fa fa-check"></i>3.2.4. Running Deblur</a></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#summarizing-deblur-output-2"><i class="fa fa-check"></i>3.2.5. Summarizing Deblur output</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#its-assign-taxonomy-to-asvs"><i class="fa fa-check"></i>3.3. ITS Assign taxonomy to ASVs</a>
<ul>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#run-taxonomic-classification-2"><i class="fa fa-check"></i>3.3.1. Run taxonomic classification</a></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#assess-subset-of-taxonomic-assignments-with-blast-2"><i class="fa fa-check"></i>3.3.2. Assess subset of taxonomic assignments with BLAST</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#its-filtering-resultant-table"><i class="fa fa-check"></i>3.4. ITS Filtering resultant table</a>
<ul>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#filter-out-rare-asvs-2"><i class="fa fa-check"></i>3.4.1. Filter out rare ASVs</a></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#filter-out-contaminant-and-unclassified-asvs-2"><i class="fa fa-check"></i>3.4.2. Filter out contaminant and unclassified ASVs</a></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#optional-exclude-low-depth-samples-2"><i class="fa fa-check"></i><em>3.4.3. (Optional) Exclude low-depth samples</em></a></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#subset-and-summarize-filtered-table-2"><i class="fa fa-check"></i>3.4.4. Subset and summarize filtered table</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="its.html"><a href="its.html#its-answers"><i class="fa fa-check"></i>ITS Answers</a></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
 <center>
 <a href="https://www.cihr-irsc.gc.ca"><img src="img/sponsors/cihr.png" width=90% style="padding-bottom: 20px" alt="Canadian Institutes of Health Research logo."></a>

 <a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.webp" width=40% style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
 <a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" width=40% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Genomics logo.">
 <br><br>
 <a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
 <a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">
 <br><br>

 <li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Beginner Microbiome Analysis 2025</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="s" class="section level1 hasAnchor">
<h1>1. 16S<a href="s.html#s" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Create a directory for Module 2 inside <code>workspace</code> and create a symlink to the raw FASTQ files and the metadata file.</p>
<pre><code>cd ~/workspace
mkdir bmb_module2/ bmb_module2/16S_analysis
cd bmb_module2/16S_analysis
ln -s ~/CourseData/MIC_data/bmb_raw_data/16S_raw_data raw_data
ln -s ~/CourseData/MIC_data/bmb_raw_data/Blueberry_16S_metadata.tsv .</code></pre>
<p>You will have installed the latest version of QIIME2 in Module 1, so you can activate that environment with the command below:</p>
<pre><code>conda activate qiime2-amplicon-2025.4</code></pre>
<p>If you didn’t get to that part, you can use this environment instead:</p>
<pre><code>conda activate qiime2-amplicon-backup</code></pre>
<p>Throughout this module, there are some questions aimed to help your understanding of some of the key concepts. You’ll find the answers at the bottom of this page, but no one will be marking them.</p>
<div id="s-first-steps" class="section level2 hasAnchor">
<h2>1.1. 16S First steps<a href="s.html#s-first-steps" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="inspect-raw-data" class="section level3 hasAnchor">
<h3>1.1.1. Inspect raw data<a href="s.html#inspect-raw-data" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>First, let’s take a look at the directory containing our raw reads as well as our metadata file.</p>
<pre><code>ls raw_data</code></pre>
<pre><code>head Blueberry_16S_metadata.tsv</code></pre>
<p><strong>Question 1: How many samples are there?</strong></p>
<p><strong>Question 2: Into what group(s) are the samples classified?</strong></p>
</div>
<div id="import-fastqs-as-qiime2-artifact" class="section level3 hasAnchor">
<h3>1.1.2. Import FASTQs as QIIME2 artifact<a href="s.html#import-fastqs-as-qiime2-artifact" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>To standardize QIIME 2 analyses and to keep track of provenance (i.e. a list of what commands were previously run to produce a file) a special format is used for all QIIME 2 input and output files called an “artifact” (with the extension QZA). The first step is to import the raw reads as a QZA file. We will first create a new directory.</p>
<pre><code>mkdir reads_qza</code></pre>
<pre><code>qiime tools import \
  --type SampleData[PairedEndSequencesWithQuality] \
  --input-path raw_data/ \
  --output-path reads_qza/reads.qza \
  --input-format CasavaOneEightSingleLanePerSampleDirFmt</code></pre>
<p>This might take a minute! If it hasn’t come back up with the command prompt that looks something like <code>(qiime2-amplicon-2025.4) ubuntu@ip-10-0-1-248:~/workspace/bmb_module2/16S_analysis$</code> yet, then it hasn’t finished running yet and you’ll need to be patient :)</p>
<p>All of the FASTQs are now in the single artifact file reads_qza/reads.qza. This file format can be a little confusing at first, but it is actually just a zipped folder. You can manipulate and explore these files better with the <code>qiime tools utilities</code> (e.g. <code>peek</code> and <code>view</code>).</p>
</div>
<div id="trim-primers-with-cutadapt" class="section level3 hasAnchor">
<h3>1.1.3. Trim primers with cutadapt<a href="s.html#trim-primers-with-cutadapt" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Screen out reads that do not begin with primer sequence and remove primer sequence from reads using the <a href="http://cutadapt.readthedocs.io/en/stable/guide.html">cutadapt</a> QIIME 2 plugin. The below primers correspond to the 16S V6-V8 region (bacteria-specific primer set). You can see more about different primers and the taxa that they target <a href="https://imr.bio/protocols.html">here</a>.</p>
<pre><code>qiime cutadapt trim-paired \
  --i-demultiplexed-sequences reads_qza/reads.qza \
  --p-cores 4 \
  --p-front-f ACGCGHNRAACCTTACC \
  --p-front-r ACGGGCRGTGWGTRCAA \
  --p-discard-untrimmed \
  --p-no-indels \
  --o-trimmed-sequences reads_qza/reads_trimmed.qza</code></pre>
<p>Visualizing your output data is a good idea after any step to make sure nothing unexpected occurred. The following command generates a “visualization” file with the extension QZV.</p>
<p>We still have a few preprocessing requirements to check off our list before denoising, so we can wait until these steps are complete to visualize our data. However, if you would like to see what paired-end reads look like before joining, run the following command and open the QZV file in <a href="https://view.qiime2.org/">QIIME2 View</a>.</p>
<pre><code>qiime demux summarize \
  --i-data reads_qza/reads_trimmed.qza \
  --o-visualization reads_qza/reads_trimmed_summary.qzv</code></pre>
<p>Remember that you can see and download these files on the webpage: <a href="http://##.uhn-hpc.ca/" class="uri">http://##.uhn-hpc.ca/</a> (replace ## with your number!)</p>
<p><strong>Question 3: What would happen if you ran this exact command on V4/V5-amplified sequences?</strong></p>
</div>
</div>
<div id="s-denoising-the-reads-into-amplicon-sequence-variants" class="section level2 hasAnchor">
<h2>1.2. 16S Denoising the reads into amplicon sequence variants<a href="s.html#s-denoising-the-reads-into-amplicon-sequence-variants" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Different denoising tools require different levels of preprocessing before the actual denoising happens. For example, DADA2 performs read joining and quality filtering as part of the denoising step itself, and can be run directly after trimming the primers. Due to speed considerations, we’ll be using Deblur instead, which requires that these steps be carried out separately. Guidelines for running DADA2 can be found <a href="https://github.com/LangilleLab/microbiome_helper/wiki/QIIME2-DADA2-Quick-Reference">here</a>.</p>
<div id="join-paired-end-reads" class="section level3 hasAnchor">
<h3>1.2.1. Join paired-end reads<a href="s.html#join-paired-end-reads" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Forward and reverse reads can be joined with <a href="https://github.com/torognes/vsearch">VSEARCH</a> as shown below. This will generate QZA files for both the joined/merged sequences and unmerged sequences.</p>
<pre><code>qiime vsearch merge-pairs \
  --i-demultiplexed-seqs reads_qza/reads_trimmed.qza \
  --output-dir reads_qza/reads_joined</code></pre>
</div>
<div id="filter-out-low-quality-reads" class="section level3 hasAnchor">
<h3>1.2.2. Filter out low-quality reads<a href="s.html#filter-out-low-quality-reads" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This command will filter out low-quality reads based on the default options.</p>
<pre><code>qiime quality-filter q-score \
  --i-demux reads_qza/reads_joined/merged_sequences.qza \
  --o-filter-stats filt_stats.qza \
  --o-filtered-sequences reads_qza/reads_trimmed_joined_filt.qza</code></pre>
</div>
<div id="summarize-joined-and-filtered-reads" class="section level3 hasAnchor">
<h3>1.2.3. Summarize joined and filtered reads<a href="s.html#summarize-joined-and-filtered-reads" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>It is a good idea at this point just to verify that there haven’t been any substantial losses of reads, before going through the whole ASV process, at either the joining or quality-filtering steps above. You will also need to select a length to trim back to that maintains the largest/acceptable quantity of reads during denoising.</p>
<pre><code>qiime demux summarize \
  --i-data reads_qza/reads_trimmed_joined_filt.qza \
  --o-visualization reads_qza/reads_trimmed_joined_filt_summary.qzv</code></pre>
<p>Now open the file in <a href="https://view.qiime2.org/">QIIME2 View</a> and look at the Overview and Interactive Quality Plot tabs to explore your data and answer the following questions.</p>
<p><strong>Question 4</strong>: How long are our forward reads? Why are there no reverse reads in our file?</p>
<p><strong>Question 5</strong>: What would be a good trim length for our reads? Remember that there are answers at the bottom of the page if you would like to check this.</p>
</div>
<div id="running-deblur" class="section level3 hasAnchor">
<h3>1.2.4. Running Deblur<a href="s.html#running-deblur" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Running the <a href="https://github.com/biocore/deblur">Deblur</a> workflow will correct the raw reads into amplicon sequence variants (ASVs). This denoising tool filters out reads that either do match to known noise or that do not match with low similarity to the expected amplicon region. Note that the below command will retain singletons, which would have been filtered out unless we set –p-min-reads 1, and is for 16S sequences only. For other amplicon regions, you can either use the denoise-other option in the command and specify a reference database of sequences to use for positive filtering (as in the below versions for 18S and ITS) or use DADA2.</p>
<p>You’ll need to replace X here with the trim length that was decided in Question 5.</p>
<pre><code>qiime deblur denoise-16S \
  --i-demultiplexed-seqs reads_qza/reads_trimmed_joined_filt.qza \
  --p-trim-length X \
  --p-sample-stats \
  --p-jobs-to-start 4 \
  --p-min-reads 1 \
  --output-dir deblur_output</code></pre>
<p><em>Note: this command may take a few minutes to run.</em></p>
</div>
<div id="summarizing-deblur-output" class="section level3 hasAnchor">
<h3>1.2.5. Summarizing Deblur output<a href="s.html#summarizing-deblur-output" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Once a denoising pipeline has been run you can summarize the output table with the below command, which will create a visualization artifact for you to view. We will use this visualization later to determine the the cut-offs for filtering the table below, but for now you should mainly take a look at the visualization to ensure that sufficient reads have been retained after running deblur. This denoising tool filters out reads that either do match to known noise or that do not match with low similarity to the expected amplicon region. If your samples have very low depth after running deblur (compared to the input read depth) this could be a red flag that either you ran the tool incorrectly, you have a lot of noise in your data, or that deblur is inappropriate for your dataset.</p>
<pre><code>qiime feature-table summarize \
  --i-table deblur_output/table.qza \
  --o-visualization deblur_output/deblur_table_summary.qzv</code></pre>
<p><strong>Question 6: What is the mean sequencing depth per sample after denoising?</strong></p>
<p><strong>Question 7: Which sample has the least reads?</strong></p>
</div>
</div>
<div id="s-assign-taxonomy-to-asvs" class="section level2 hasAnchor">
<h2>1.3. 16S Assign taxonomy to ASVs<a href="s.html#s-assign-taxonomy-to-asvs" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<pre><code>You can assign taxonomy to your ASVs using a Naive-Bayes approach implemented in the [scikit learn](http://scikit-learn.org/stable/) Python library and the [SILVA](https://www.arb-silva.de/) or [UNITE](https://unite.ut.ee/) databases. This approach requires that a classifier be trained in advance on a reference database. We recommend users use a widely used classifier to help ensure there are no unexpected issues with the Naive-Bayes model. We previously maintained primer-specific classifiers, which theoretically can provide more accurate classifications, but we no longer do this due to concerns regarding issues with the trained models that are difficult to catch if only a couple people are running them. The full-length 16S/18S classifier can be downloaded from the [QIIME 2 website](https://docs.qiime2.org/2022.11/data-resources/) (silva-138-99-nb-classifier.qza for the latest classifier). Custom classifiers for the ITS region that we have generated from the UNITE database are available as well ([see downloads](http://kronos.pharmacology.dal.ca/public_files/MH/taxa_classifiers/qiime2-2020.8_classifiers) and [commands used to create these files](https://github.com/LangilleLab/microbiome_helper/wiki/Creating-QIIME-2-Taxonomic-Classifiers)):

-   Full ITS - fungi only (classifier_sh_refs_qiime_ver9_99_s_27.10.2022_ITS.qza)
-   Full ITS - all eukaryotes (classifier_sh_refs_qiime_ver9_99_s_all_27.10.2022_ITS.qza)</code></pre>
<p>We’re going to use the Greengenes2 classifier. We can download that from the FTP site <a href="http://ftp.microbio.me/greengenes_release/2022.10/">here</a> - click on the folder <code>sklearn-1.4.2-compatible-nb-classifiers/</code>, and then copy the link for the <code>2022.10.backbone.full-length.nb.sklearn-1.4.2.qza</code> classifier. (Right-click and select “Copy Link Address”. Then we’ll use the <code>wget</code> command like we did in Module 1. That will look something like this:</p>
<pre><code>wget link-address</code></pre>
<div id="run-taxonomic-classification" class="section level3 hasAnchor">
<h3>1.3.1. Run taxonomic classification<a href="s.html#run-taxonomic-classification" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>You can run the taxonomic classification with this command, which is one of the longest running and most memory-intensive command of the tutorial. If you receive an error related to insufficient memory (and if you cannot increase your memory usage) then you can look into the –p-reads-per-batch option and set this to be lower than the default (which is dynamic depending on sample depth and the number of threads) and also try running the command with fewer jobs (e.g. set –p-n-jobs 1).</p>
<p><strong>Because of the memory constraints of this workshop, you can copy the output into your folder:</strong></p>
<p>But here is the code you would use to run the taxonomic classification:</p>
<pre><code>qiime feature-classifier classify-sklearn \
  --i-reads deblur_output/representative_sequences.qza \
  --i-classifier 2022.10.backbone.full-length.nb.sklearn-1.4.2.qza \
  --p-n-jobs 4 \
  --output-dir taxa</code></pre>
<p>Unfortunately we don’t actually have enough memory to run this so you’ll see a <code>Killed</code> output for this.
Instead, we’ll copy across the output that we would have got.</p>
<pre><code>mkdir taxa      
cp ~/CourseData/MIC_data/bmb_module2/16S/taxa/classification.qza taxa</code></pre>
<p>As with all QZA files, you can export the output file to take a look at the classifications and confidence scores:</p>
<pre><code>qiime tools export \
  --input-path taxa/classification.qza \
  --output-path taxa</code></pre>
</div>
<div id="assess-subset-of-taxonomic-assignments-with-blast" class="section level3 hasAnchor">
<h3>1.3.2 Assess subset of taxonomic assignments with BLAST<a href="s.html#assess-subset-of-taxonomic-assignments-with-blast" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>The performance of the taxonomic classification is difficult to assess without a gold-standard reference, but nonetheless one basic sanity check is to compare the taxonomic assignments with the top BLASTn hits for certain ASVs. First, generate a QZV file for the denoised representative sequences in QIIME 2 by running:</p>
<pre><code>qiime feature-table tabulate-seqs \
  --i-data deblur_output/representative_sequences.qza \
  --o-visualization deblur_output/representative_sequences.qzv</code></pre>
<p>This QZV file tabulates the denoised sequences. Clicking on the nucleotide sequence links to a BLASTn search for that sequence. By comparing these BLAST hits with the taxonomic assignment of ASVs generated above you can reassure yourself that the taxonomic assignments overall worked correctly. It’s a good idea to select ~5 ASVs to BLAST for this validation, which should be from taxonomically different groups, such as different phyla, according to the taxonomic classifier.</p>
<p>You can then check the taxonomy that was assigned to them using your classifier - you can either open up the <code>taxonomy.tsv</code> file and search for the Feature ID/ASV names there, or you can use the <code>grep</code> command like so:</p>
<pre><code>grep &quot;02fda7604175053e87dc11e2598a89e0&quot; taxa/taxonomy.tsv</code></pre>
</div>
</div>
<div id="s-filtering-resultant-table" class="section level2 hasAnchor">
<h2>1.4. 16S Filtering resultant table<a href="s.html#s-filtering-resultant-table" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Filtering the denoised table is an important step of microbiome data analysis. You can see more details on this process in the <a href="https://docs.qiime2.org/2022.11/tutorials/filtering/">QIIME 2 filtering tutorial</a>.</p>
<div id="filter-out-rare-asvs" class="section level3 hasAnchor">
<h3>1.4.1. Filter out rare ASVs<a href="s.html#filter-out-rare-asvs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Based on the summary visualization created in step 2.5 above you can choose a cut-off for how frequent a variant needs to be (and optionally how many samples need to have the variant) for it to be retained. Here we will remove all ASVs that have a frequency of less than 0.1% of the mean sample depth. This cut-off excludes ASVs that are likely due to MiSeq bleed-through between runs (reported by Illumina to be 0.1% of reads). To calculate this cut-off you would identify the mean sample depth in the visualization created in step 2.5 (<code>deblur_table_summary.qzv</code>), multiply it by 0.001, and round to the nearest integer.</p>
<p>Once you’ve determined how you would like to filter your table you can do so with this command (X is a placeholder for your choice):</p>
<pre><code>qiime feature-table filter-features \
  --i-table deblur_output/table.qza \
  --p-min-frequency X \
  --p-min-samples 1 \
  --o-filtered-table deblur_output/deblur_table_filt.qza</code></pre>
</div>
<div id="filter-out-contaminant-and-unclassified-asvs" class="section level3 hasAnchor">
<h3>1.4.2. Filter out contaminant and unclassified ASVs<a href="s.html#filter-out-contaminant-and-unclassified-asvs" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Once we have assigned taxonomy to our ASVs we can use that information to remove ASVs which are likely contaminants or noise based on the taxonomic labels. Two common contaminants in 16S sequencing data are mitochondrial and chloroplast 16S sequences, which can be removed by excluding any ASV which contains those terms in its taxonomic label. It can also be <strong>sometimes</strong> useful to exclude any ASV that is unclassified at the phylum level since these sequences could be noise (e.g. possible chimeric sequences). Note that if your data has not been classified against the default database you may need to change p__ to be a string that enables phylum-level assignments to be identified or simply omit that line.</p>
<p>In general though, it can be very informative if your sequencing reads are coming back with significant amounts of unclassified ASVs as it can indicate upstream analysis problems or indicate you are studying a poorly characterized environment where you have a good chance of identifying a lot of novel phyla. Therefore, our recommendation is to not filter out the unclassified sequences by default, but we will do so here.</p>
<pre><code>qiime taxa filter-table \
  --i-table deblur_output/deblur_table_filt.qza \
  --i-taxonomy taxa/classification.qza \
  --p-include p__ \
  --p-exclude mitochondria,chloroplast \
  --o-filtered-table deblur_output/deblur_table_filt_contam.qza</code></pre>
</div>
<div id="optional-exclude-low-depth-samples" class="section level3 hasAnchor">
<h3><em>1.4.3. (Optional) Exclude low-depth samples</em><a href="s.html#optional-exclude-low-depth-samples" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Often certain samples will have quite low depth after these filtering steps, which can be excluded from downstream analyses since they will largely add noise. There is no single cut-off that works best for all datasets, but researchers often use minimum cut-offs within the range of 1000 to 4000 reads. You can also use a cut-off much lower than this if you want to retain all samples except those that failed entirely (e.g. depth &lt; 50 reads).</p>
<p>Ideally you would choose this cut-off after visualizing rarefaction curves to determine at what read depth the richness of your samples plateaus and choose a cut-off as close to this plateau as possible while retaining sufficient sample size for your analyses. We learn more about rarefaction curves in Module 3.</p>
</div>
<div id="subset-and-summarize-filtered-table" class="section level3 hasAnchor">
<h3>1.4.4. Subset and summarize filtered table<a href="s.html#subset-and-summarize-filtered-table" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Check output after filtering.</p>
<pre><code>qiime feature-table summarize \
  --i-table deblur_output/deblur_table_filt_contam.qza \
  --o-visualization deblur_output/deblur_table_filt_contam_summary.qzv</code></pre>
<p><strong>Question 8: What is the minimum and maximum sequencing depth across all samples?</strong></p>
<p>Happy? Copy a final table.</p>
<pre><code>cp deblur_output/deblur_table_filt_contam.qza deblur_output/deblur_table_final.qza</code></pre>
<p>Once we have our final filtered table we will need to subset the QZA file containing the ASV sequences to the same set. You can exclude any removed ASVs from the sequence file with this command:</p>
<pre><code>qiime feature-table filter-seqs \
  --i-data deblur_output/representative_sequences.qza \
  --i-table deblur_output/deblur_table_final.qza  \
  --o-filtered-data deblur_output/rep_seqs_final.qza</code></pre>
<p>Finally, you can make a new summary of the final filtered abundance table:</p>
<pre><code>qiime feature-table summarize \
  --i-table deblur_output/deblur_table_final.qza \
  --o-visualization deblur_output/deblur_table_final_summary.qzv</code></pre>
</div>
</div>
<div id="s-answers" class="section level2 hasAnchor">
<h2>16S Answers<a href="s.html#s-answers" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p><strong>Question 1: How many samples are there?</strong></p>
<p>We have 10 samples. The raw data folder contains one fastq.gz file for each set of the forward and reverse sequences (labelled R1 and R2, respectively) for each of the samples (B-Rtxxx).</p>
<p><strong>Question 2: Into what group(s) are the samples classified?</strong></p>
<p>The samples we are using are classified into two different groups: Forest or Managed sites. The study we’re looking at does also look at bulk or rhizosphere samples, but we’re just using a small subset of the samples for this workshop.</p>
<p><strong>Question 3: What would happen if you ran this exact command on V4/V5-amplified sequences?</strong></p>
<p><code>Cutadapt</code> will only trim reads that match the specified primer sequence. Therefore, most reads would be discarded because we are including the <code>--p-discard-untrimmed</code> option.</p>
<p><strong>Question 4: How long are our forward reads? Why are there no reverse reads in our file?</strong></p>
<p>The forward read median length is 405 nucleotides. There are no reverse reads because forward and reverse reads were merged into one sequence during read joining.</p>
<p><strong>Question 5: What would be a good trim length for our reads?</strong></p>
<p>There is no one right answer for this question, but a trim length of 390 nucleotides will maintain most of our sequences apart from any that are really short. You could choose a different length, but this will give different answers further on. Typically, this choice is a trade-off between maintaining a longer sequence length where we’re more likely to get finer-resolution taxonomic classification, or trimming more so that we retain more sequences.</p>
<p><strong>Question 6: What is the mean sequencing depth per sample after denoising?</strong></p>
<p>The mean sequencing depth (frequency) across all denoised samples is 6,669.8 reads.</p>
<p><strong>Question 7: Which sample has the least reads?</strong></p>
<p>Sample B-Rt151 has the lowest sequencing depth (4,332 reads).</p>
<p><strong>Question 8: What is the minimum and maximum sequencing depth across all samples?</strong></p>
<p>The final minimum sequencing depth is 3,432 and the maximum sequencing depth is 8,650 reads.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-2-lab-marker-gene-profiling.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="s-1.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": true,
    "twitter": true,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/cbw-dev/bookdown-template/blob/main/030-module-2.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

</body>

</html>
